From 9f85a145810434fbdc81178072fb097be801ccb6 Mon Sep 17 00:00:00 2001
From: Guochun Huang <hero.huang@rock-chips.com>
Date: Fri, 8 Dec 2023 09:31:56 +0000
Subject: [PATCH 188/232] misc: rk628: hdmitx: add color_bar debugfs node

Enable normal color bar
echo 1 > /sys/kernel/debug/rk628/2-0050/hdmitx_color_ba

Enable special color bar
echo 2 > /sys/kernel/debug/rk628/2-0050/hdmitx_color_bar

Enable black color bar
echo 3 > /sys/kernel/debug/rk628/2-0050/hdmitx_color_bar

Disable color bar
echo 0 > /sys/kernel/debug/rk628/2-0050/hdmitx_color_bar

Change-Id: Ie2a17a43bc2bb112d8a9653e87a86cb2c80bb67a
Signed-off-by: Guochun Huang <hero.huang@rock-chips.com>
---
 drivers/misc/rk628/rk628.c        |  1 +
 drivers/misc/rk628/rk628_hdmitx.c | 72 +++++++++++++++++++++++++++++++
 drivers/misc/rk628/rk628_hdmitx.h | 10 ++++-
 3 files changed, 82 insertions(+), 1 deletion(-)

diff --git a/drivers/misc/rk628/rk628.c b/drivers/misc/rk628/rk628.c
index a436100692eaf9..d70e777269ba54 100644
--- a/drivers/misc/rk628/rk628.c
+++ b/drivers/misc/rk628/rk628.c
@@ -1225,6 +1225,7 @@ static void rk628_debugfs_create(struct rk628 *rk628)
 	rk628_post_process_create_debugfs_file(rk628);
 	rk628_mipi_dsi_create_debugfs_file(rk628);
 	rk628_gvi_create_debugfs_file(rk628);
+	rk628_hdmitx_create_debugfs_file(rk628);
 }
 
 static int
diff --git a/drivers/misc/rk628/rk628_hdmitx.c b/drivers/misc/rk628/rk628_hdmitx.c
index 1f3c2c12075803..8949b3a9a9d9bc 100644
--- a/drivers/misc/rk628/rk628_hdmitx.c
+++ b/drivers/misc/rk628/rk628_hdmitx.c
@@ -1480,6 +1480,78 @@ static void rk628_hdmi_cec_unregister(struct rk628_hdmi_cec *cec)
 	cec_unregister_adapter(cec->adap);
 }
 
+static int rk628_hdmitx_color_bar_show(struct seq_file *s, void *data)
+{
+	seq_puts(s, "  Enable normal color bar:\n");
+	seq_puts(s, "      example: echo 1 > /sys/kernel/debug/rk628/2-0050/hdmitx_color_bar\n");
+	seq_puts(s, "  Enable special color bar:\n");
+	seq_puts(s, "      example: echo 2 > /sys/kernel/debug/rk628/2-0050/hdmitx_color_bar\n");
+	seq_puts(s, "  Enable black color bar:\n");
+	seq_puts(s, "      example: echo 3 > /sys/kernel/debug/rk628/2-0050/hdmitx_color_bar\n");
+	seq_puts(s, "  Disable color bar:\n");
+	seq_puts(s, "      example: echo 0 > /sys/kernel/debug/rk628/2-0050/hdmitx_color_bar\n");
+
+	return 0;
+}
+
+static int rk628_hdmitx_color_bar_open(struct inode *inode, struct file *file)
+{
+	return single_open(file, rk628_hdmitx_color_bar_show, inode->i_private);
+}
+
+static ssize_t rk628_hdmitx_color_bar_write(struct file *file, const char __user *ubuf,
+					    size_t len, loff_t *offp)
+{
+	struct rk628 *rk628 = ((struct seq_file *)file->private_data)->private;
+	u8 mode;
+
+	if (kstrtou8_from_user(ubuf, len, 0, &mode))
+		return -EFAULT;
+
+	switch (mode) {
+	case 0:
+		rk628_i2c_update_bits(rk628, HDMI_COLOR_BAR, DISABLE_COLORBAR_BIST_MASK,
+				      DISABLE_COLORBAR_BIST(1));
+		break;
+	case 1:
+		rk628_i2c_update_bits(rk628, HDMI_COLOR_BAR, DISABLE_COLORBAR_BIST_MASK,
+				      DISABLE_COLORBAR_BIST(0));
+		rk628_i2c_update_bits(rk628, HDMI_COLOR_BAR, VIDEO_BIST_MODE_MASK,
+				      VIDEO_BIST_MODE(0));
+		break;
+	case 2:
+		rk628_i2c_update_bits(rk628, HDMI_COLOR_BAR, DISABLE_COLORBAR_BIST_MASK,
+				      DISABLE_COLORBAR_BIST(0));
+		rk628_i2c_update_bits(rk628, HDMI_COLOR_BAR, VIDEO_BIST_MODE_MASK,
+				      VIDEO_BIST_MODE(1));
+		break;
+	case 3:
+	default:
+		rk628_i2c_update_bits(rk628, HDMI_COLOR_BAR, DISABLE_COLORBAR_BIST_MASK,
+				      DISABLE_COLORBAR_BIST(0));
+		rk628_i2c_update_bits(rk628, HDMI_COLOR_BAR, VIDEO_BIST_MODE_MASK,
+				      VIDEO_BIST_MODE(2));
+	}
+
+	return len;
+}
+
+static const struct file_operations rk628_hdmitx_color_bar_fops = {
+	.owner = THIS_MODULE,
+	.open = rk628_hdmitx_color_bar_open,
+	.read = seq_read,
+	.write = rk628_hdmitx_color_bar_write,
+	.llseek = seq_lseek,
+	.release = single_release,
+};
+
+void rk628_hdmitx_create_debugfs_file(struct rk628 *rk628)
+{
+	if (rk628_output_is_hdmi(rk628))
+		debugfs_create_file("hdmitx_color_bar", 0600, rk628->debug_dir,
+				    rk628, &rk628_hdmitx_color_bar_fops);
+}
+
 void rk628_hdmitx_disable(struct rk628 *rk628)
 {
 	rk628_i2c_update_bits(rk628, HDMI_SYS_CTRL, POWER_MASK, PWR_OFF(1));
diff --git a/drivers/misc/rk628/rk628_hdmitx.h b/drivers/misc/rk628/rk628_hdmitx.h
index c474182c94cd2f..da97e441a0c890 100644
--- a/drivers/misc/rk628/rk628_hdmitx.h
+++ b/drivers/misc/rk628/rk628_hdmitx.h
@@ -342,13 +342,18 @@ enum {
 #define HDMI_CEC_BUSFREETIME_L		HDMITX_REG(0xdc)
 #define HDMI_CEC_BUSFREETIME_H		HDMITX_REG(0xdd)
 #define HDMI_CEC_LOGICADDR		HDMITX_REG(0xde)
-
+#define HDMI_COLOR_BAR			HDMITX_REG(0xc9)
+#define VIDEO_BIST_MODE_MASK		GENMASK(7, 6)
+#define VIDEO_BIST_MODE(x)		UPDATE(x, 7, 6)
+#define DISABLE_COLORBAR_BIST_MASK	BIT(4)
+#define DISABLE_COLORBAR_BIST(x)	UPDATE(x, 4, 4)
 #define HDMI_MAX_REG			HDMITX_REG(0xed)
 
 #ifdef CONFIG_DRM
 void rk628_hdmitx_disable(struct rk628 *rk628);
 int rk628_hdmitx_enable(struct rk628 *rk628);
 void rk628_hdmitx_remove(struct rk628 *rk628);
+void rk628_hdmitx_create_debugfs_file(struct rk628 *rk628);
 #else
 static inline void rk628_hdmitx_disable(struct rk628 *rk628)
 {
@@ -361,7 +366,10 @@ static inline int rk628_hdmitx_enable(struct rk628 *rk628)
 
 static inline void rk628_hdmitx_remove(struct rk628 *rk628)
 {
+}
 
+static inline void rk628_hdmitx_create_debugfs_file(struct rk628 *rk628)
+{
 }
 #endif
 
-- 
2.17.1

